File: hamersapp/src/main/java/nl/ecci/hamers/quotes/QuoteFragment.java
package nl.ecci.hamers.quotes;

import android.content.DialogInterface;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.view.MenuItemCompat;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SearchView;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import com.android.volley.VolleyError;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.ArrayList;

import nl.ecci.hamers.MainActivity;
import nl.ecci.hamers.R;
import nl.ecci.hamers.helpers.DividerItemDecoration;
import nl.ecci.hamers.helpers.HamersFragment;
import nl.ecci.hamers.loader.GetCallback;
import nl.ecci.hamers.loader.Loader;

import static nl.ecci.hamers.MainActivity.prefs;

public class QuoteFragment extends HamersFragment implements DialogInterface.OnDismissListener {

    private final ArrayList<Quote> dataSet = new ArrayList<>();
    private QuoteAdapter adapter;
    private RecyclerView quote_list;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.hamers_fragment, container, false);
        quote_list = (RecyclerView) view.findViewById(R.id.hamers_recyclerview);

        setHasOptionsMenu(true);

        LinearLayoutManager mLayoutManager = new LinearLayoutManager(getActivity());
        quote_list.setLayoutManager(mLayoutManager);
        quote_list.addItemDecoration(new DividerItemDecoration(getActivity()));

        adapter = new QuoteAdapter(dataSet, getContext());
        quote_list.setAdapter(adapter);

        swipeRefreshLayout = (SwipeRefreshLayout) view.findViewById(R.id.hamers_swipe_container);
        initSwiper(quote_list, mLayoutManager, swipeRefreshLayout);

        // When user presses "+" in QuoteListFragment, start new dialog with NewQuoteFragment
        final FloatingActionButton newQuoteButton = (FloatingActionButton) view.findViewById(R.id.hamers_fab);
        newQuoteButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                NewQuoteFragment newQuoteFragment = new NewQuoteFragment();
                newQuoteFragment.show(getChildFragmentManager(), "quotes");
            }
        });

        new populateList().execute();
        onRefresh();

        return view;
    }

    @Override
    public void onRefresh() {
        setRefreshing(true);
        Loader.getData(getContext(), Loader.QUOTEURL, new GetCallback() {
            @Override
            public void onSuccess(String response) {
                new populateList().execute(response);
            }

            @Override
            public void onError(VolleyError error) {
                // Nothing
            }
        }, null);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.quote_menu, menu);
        MenuItem menuItem = menu.findItem(R.id.quote_search);
        SearchView searchView = (SearchView) MenuItemCompat.getActionView(menuItem);
        if (searchView != null) {
            searchView.setQueryHint(getString(R.string.search_hint));
            searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
                @Override
                public boolean onQueryTextSubmit(String s) {
                    return false;
                }

                @Override
                public boolean onQueryTextChange(String s) {
                    adapter.getFilter().filter(s.toLowerCase());
                    return false;
                }
            });
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        onRefresh();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.scroll_top:
                scrollTop();
                return true;
            default:
                return false;
        }
    }

    private void scrollTop() {
        quote_list.smoothScrollToPosition(0);
    }

    @Override
    public void onDismiss(DialogInterface dialogInterface) {
        onRefresh();
    }

    private class populateList extends AsyncTask<String, Void, ArrayList<Quote>> {

        @Override
        protected final ArrayList<Quote> doInBackground(String... params) {
            ArrayList<Quote> result;
            GsonBuilder gsonBuilder = new GsonBuilder();
            gsonBuilder.setDateFormat(MainActivity.dbDF.toPattern());
            Gson gson = gsonBuilder.create();
            Type type = new TypeToken<ArrayList<Quote>>() {
            }.getType();

            if (params.length > 0) {
                result = gson.fromJson(params[0], type);
            } else {
                result = gson.fromJson(prefs.getString(Loader.QUOTEURL, null), type);
            }
            return result;
        }

        @Override
        protected void onPostExecute(ArrayList<Quote> result) {
            if (result != null) {
                dataSet.clear();
                dataSet.addAll(result);
                if (adapter != null) {
                    adapter.notifyDataSetChanged();
                }
            }
            setRefreshing(false);
        }
    }
}

File: hamersapp/src/main/java/nl/ecci/hamers/quotes/QuoteFragment.kt
package nl.ecci.hamers.quotes

import android.content.DialogInterface
import android.os.AsyncTask
import android.os.Bundle
import android.support.v4.view.MenuItemCompat
import android.support.v7.widget.SearchView
import android.view.*
import com.google.gson.GsonBuilder
import com.google.gson.reflect.TypeToken
import kotlinx.android.synthetic.main.fragment_hamers_list.*
import kotlinx.android.synthetic.main.fragment_hamers_list.view.*
import nl.ecci.hamers.MainActivity
import nl.ecci.hamers.MainActivity.prefs
import nl.ecci.hamers.R
import nl.ecci.hamers.helpers.DividerItemDecoration
import nl.ecci.hamers.helpers.HamersListFragment
import nl.ecci.hamers.loader.GetCallback
import nl.ecci.hamers.loader.Loader
import org.jetbrains.anko.padding
import java.util.*


class QuoteFragment : HamersListFragment(), DialogInterface.OnDismissListener {

    private val dataSet = ArrayList<Quote>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        retainInstance = true
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        setHasOptionsMenu(true)
        return inflater?.inflate(R.layout.fragment_hamers_list, container, false)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        hamers_list.addItemDecoration(DividerItemDecoration(activity))
        hamers_list.padding = 0
        hamers_list.adapter = QuoteAdapter(dataSet, context)

        // When user presses "+" in QuoteListFragment, start new dialog with NewQuoteFragment
        hamers_fab.setOnClickListener {
            NewQuoteFragment().show(childFragmentManager, "quotes")
        }

        populateList().execute()
    }

    override fun onRefresh() {
        setRefreshing(true)
        Loader.getData(context, Loader.QUOTEURL, object : GetCallback {
            override fun onSuccess(response: String) {
                populateList().execute(response)
            }
        }, null)
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.quote_menu, menu)
        val menuItem = menu.findItem(R.id.quote_search)
        val searchView = MenuItemCompat.getActionView(menuItem) as SearchView
        searchView.queryHint = getString(R.string.search_hint)
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(s: String): Boolean {
                return false
            }

            override fun onQueryTextChange(s: String): Boolean {
                (hamers_list.adapter as QuoteAdapter).filter.filter(s.toLowerCase())
                return false
            }
        })
    }

    override fun onResume() {
        super.onResume()
        onRefresh()
        activity.title = resources.getString(R.string.navigation_item_quotes)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.scroll_top -> {
                scrollTop()
                return true
            }
            else -> return false
        }
    }

    private fun scrollTop() {
        view?.hamers_list?.smoothScrollToPosition(0)
    }

    override fun onDismiss(dialogInterface: DialogInterface) {
        onRefresh()
    }

    private inner class populateList : AsyncTask<String, Void, ArrayList<Quote>>() {

        override fun doInBackground(vararg params: String): ArrayList<Quote> {
            val result: ArrayList<Quote> = ArrayList()
            val gsonBuilder = GsonBuilder()
            gsonBuilder.setDateFormat(MainActivity.dbDF.toPattern())
            val gson = gsonBuilder.create()
            val type = object : TypeToken<ArrayList<Quote>>() {

            }.type

            val quotes = prefs.getString(Loader.QUOTEURL, null)

            if (params.isNotEmpty()) {
                result.addAll(gson.fromJson<ArrayList<Quote>>(params[0], type))
            } else if (quotes != null) {
                result.addAll(gson.fromJson<ArrayList<Quote>>(quotes, type))
            }
            return result
        }

        override fun onPostExecute(result: ArrayList<Quote>?) {
            if (result != null) {
                dataSet.clear()
                dataSet.addAll(result)
                notifyAdapter()
            }
            setRefreshing(false)
        }
    }
}